examples/13_regions.cpp

```diff
@@ -0,0 +1,261 @@
+       << "    y: " << y << "\n";
+
+  cout << "  Arithmetic operations:\n";
+  auto z = x + 2 * y;
+  cout << "    z: " << z << "\n";
+
+  cout << "  Unit vectors pointing in direction d:\n";
+  auto u0 = Point<int, 2>::unit(0);
+  auto u1 = Point<int, 2>::unit(1);
+  cout << "    u0: " << u0 << "\n"
+       << "    u1: " << u1 << "\n";
+  cout << "  A points with all elements the same:\n";
+  auto p3 = Point<int, 2>::pure(3);
+  cout << "    p3: " << p3 << "\n";
+
+  cout << "  Element-wise operations:\n";
+  using std::abs, std::max;
+  auto mxy1 = max(abs(x), abs(y));
+  // Apply arbitrary functions element-wise
+  auto mxy2 = fmap([](auto a, auto b) { return max(abs(a), abs(b)); }, x, y);
+  cout << "    mxy1: " << mxy1 << "\n"
+       << "    mxy2: " << mxy2 << "\n";
+
+  cout << "  Reduction operations:\n";
+  int mx1 = max_element(x);
+  // Apply arbitrary reduction operations
+  int mx2 = fold([](auto r, auto a) { return max(r, a); }, 0, x);
+  cout << "    mx1: " << mx1 << "\n"
+       << "    mx2: " << mx2 << "\n";
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+void ndpoint_example() {
+  using namespace openPMD::Regions;
+  using namespace std;
+
+  cout << "\n"
+       << "NDPoints (with dimension only known at run time):\n";
+
+  cout << "  Define two points:\n";
+  NDPoint<int> x{1, 2}, y{4, 5};
+  cout << "    x: " << x << "\n"
+       << "    y: " << y << "\n";
+
+  // Arithmetic operations
+  cout << "  Arithmetic operations:\n";
+  auto z = x + 2 * y;
+  cout << "    z: " << z << "\n";
+
+  cout << "  Unit vectors pointing in direction d:\n";
+  auto u0 = NDPoint<int>::unit(2, 0);
+  auto u1 = NDPoint<int>::unit(2, 1);
+  cout << "    u0: " << u0 << "\n"
+       << "    u1: " << u1 << "\n";
+  cout << "  A points with all elements the same:\n";
+  auto p3 = NDPoint<int>::pure(2, 3);
+  cout << "    p3: " << p3 << "\n";
+
+  cout << "  Element-wise operations:\n";
+  using std::abs, std::max;
+  auto mxy1 = max(abs(x), abs(y));
+  // Apply arbitrary functions element-wise
+  auto mxy2 = fmap([](auto a, auto b) { return max(abs(a), abs(b)); }, x, y);
+  cout << "    mxy1: " << mxy1 << "\n"
+       << "    mxy2: " << mxy2 << "\n";
+
+  cout << "  Reduction operations:\n";
+  int mx1 = max_element(x);
+  // Apply arbitrary reduction operations
+  int mx2 = fold([](auto r, auto a) { return max(r, a); }, 0, x);
+  cout << "    mx1: " << mx1 << "\n"
+       << "    mx2: " << mx2 << "\n";
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+void box_example() {
+  using namespace openPMD::Regions;
+  using namespace std;
+
+  cout << "\n"
+       << "Boxes are spanned between points (inclusive lower, exclusive upper "
+          "bound):\n";
+
+  cout << "  Define two points:\n";
+  Point<int, 2> x{1, 4}, y{2, 5};
+  cout << "  x:" << x << "\n"
+       << "  y:" << y << "\n";
+
+  cout << "  Define a box between these points:\n";
+  Box<int, 2> b(x, y);
```

Comment:
This looks all very nice.

Can we construct a `Point` with an `openPMD::Offset` and `openPMD::Extent` (and then build a `Box` from it)?
At the moment, offset/extent are just aliases to `std::vector`s but should be equivalent to points.

Reply:
Yes, there are constructors from and conversion operators to `std::vector`. If the dimension is known at compile time, the vector length must be the same (`Point`). Otherwise (`NDPoint`), the dimension is taken from the vector length.